apps/aecore/src/aec_chain_state.erl

```diff
@@ -721,51 +783,91 @@ assert_state_hash_valid(Trees, Node) ->
-find_prev_key_nodes(Node, N) when N > 0 ->
-    %% Assert
-    key = node_type(Node),
-    find_prev_key_nodes({ok, Node}, N, none).
-
-find_prev_key_nodes(error,_N,_Acc) ->
-    error;
-find_prev_key_nodes({ok, Node}, 0, Acc) ->
-    {Node, Acc};
-find_prev_key_nodes({ok, Node}, 1,_Acc) ->
-    find_prev_key_nodes(db_find_node(prev_key_hash(Node)), 0, Node);
-find_prev_key_nodes({ok, Node}, N, Acc) ->
-    find_prev_key_nodes(db_find_node(prev_key_hash(Node)), N - 1, Acc).
-
-grant_fees(Node, Trees, Delay, Fees, State) ->
-    {KeyNode1, KeyNode2} = find_prev_key_nodes(Node, Delay + 1),
-    KeyFees = case Node =:= KeyNode2 of
-                  true  -> Fees;
-                  false -> db_get_fees(hash(KeyNode2))
-              end,
+find_predecessor_at_height(Node, Height) ->
+    case node_height(Node) of
+        Height -> Node;
+        H when H < Height -> error({cannot_find_predecessor, H, Height});
+        H ->
+            case db_find_key_nodes_at_height(Height) of
+                {ok, [KeyNode]} -> KeyNode;
+                {ok, KeyNodes} -> find_one_predecessor(KeyNodes, Node);
+                error -> error({cannot_find_predecessor, H, Height})
+            end
+    end.
+
+find_one_predecessor([N|Left], Node) ->
+    Hash1 = hash(N),
+    Hash2 = hash(Node),
+    case find_fork_point(Hash1, Hash2) of
+        {ok, Hash1} -> N;
+        {ok, _} -> find_one_predecessor(Left, Node);
+        error -> find_one_predecessor(Left, Node)
+    end.
+
+grant_fees(Node, Trees, Delay, FraudStatus, State) ->
+    KeyNode4 = find_predecessor_at_height(Node, node_height(Node) - Delay + 2),
+    KeyNode3 = db_get_node(prev_key_hash(KeyNode4)),
+    KeyNode2 = db_get_node(prev_key_hash(KeyNode3)),
+    KeyNode1 = db_get_node(prev_key_hash(KeyNode2)),
+    FraudStatus1 = db_get_fraud_status(hash(KeyNode3)),
+    FraudStatus2 = case KeyNode4 =:= Node of
+                       true  -> FraudStatus;
+                       false -> db_get_fraud_status(hash(KeyNode4))
+                   end,
+    KeyFees  = db_get_fees(hash(KeyNode2)),
     Beneficiary1 = node_beneficiary(KeyNode1),
     Beneficiary2 = node_beneficiary(KeyNode2),
     Reward = aec_governance:block_mine_reward(),
+    FraudReward = aec_governance:fraud_report_reward(),
     BeneficiaryReward1 = round(KeyFees * 0.4),
     BeneficiaryReward2 = KeyFees - BeneficiaryReward1 + Reward,
-    Trees1 = aec_trees:grant_fee(Beneficiary2, Trees, BeneficiaryReward2),
-    case node_is_genesis(KeyNode1, State) of
-        true  -> Trees1;
-        false -> aec_trees:grant_fee(Beneficiary1, Trees1, BeneficiaryReward1)
+    case {FraudStatus1, FraudStatus2} of
+        {true, true} ->
+            %% The miner of KeyNode1 was reported by the miner of KeyNode2
+            %% but that miner was reported by the miner of KeyNode3.
+            %% No rewards at all.
+            Trees;
+        {true, false} ->
+            %% The miner of KeyNode1 was reported by the miner of KeyNode2
```

Comment:
Context: fraudulent miner of KeyNode1 has already (see case `{false, true}` below) been deprived of:
* The coinbase for having mined KeyNode1; and
* The 60% of the fees in the generation that it closed with KeyNode1 (if that generation contained any txs).

Now it is being also deprived of the 40% of the fees in the microblocks (if any) it signed in the generation opened by the block it mined (KeyNode1).

Ok.